#Check if packages are downloaded
if (!require("shiny")) install.packages("shiny")
if (!require("shinydashboard")) install.packages("shinydashboard")
if (!require("progress")) install.packages("progress")
if (!require("data.table")) install.packages("data.table")
if (!require("vcfR")) install.packages("vcfR")
if (!require("DT")) install.packages("DT")

#Load packages
library(shiny)
library(shinydashboard)
library(progress)
library(data.table)
library(vcfR)
library(DT)  # For interactive tables

#Set UI function
ui <- fluidPage(
  dashboardPage(
    dashboardHeader(title = span(img(src = "rareinsight_final.png", height = 50), "RareInsight")),
    dashboardSidebar(
      sidebarMenu(
        menuItem("Home ", tabName = "home", icon = icon("home")),
        menuItem("Services", tabName = "service", icon = icon("bar-chart"),
                 menuSubItem("Input User Information", tabName = "input_user_info", icon = icon("user-circle")),
                 menuSubItem("Search Panel", tabName = "Search", icon = icon("search", lib = "glyphicon" )),
                 menuSubItem("Vcf Panel", tabName = "vcfpanel", icon = icon("file")),
                 menuItem("Diagnostic report", tabName = "diagnostic_report", icon = icon("line-chart"))
        ),
        menuItem("Acknowledgement", tabName = "acknow", icon = icon("handshake"))
      )
    ),
    dashboardBody(
      tags$head(tags$style(HTML('
                                /* logo */
                                .skin-blue .main-header .logo {
                                background-color: #030637;
                                }

                                /* navbar (rest of the header) */
                                .skin-blue .main-header .navbar {
                                background-color: #030637;
                                }

                                /* logo when hovered */
                                .skin-blue .main-header .logo:hover {
                                background-color: #030637;
                                }

                                /* main sidebar */
                                .skin-blue .main-sidebar {
                                background-color: #3C0753;
                                }

                                /* active selected tab in the sidebarmenu */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu .active a{
                                background-color: #9f16db;
                                }

                                /* other links in the sidebarmenu */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu a{
                                background-color: ;
                                color: #FFFFFF
                                }

                                /* other links in the sidebarmenu when hovered */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu a:hover{
                                background-color: #030637;
                                }
                                /* toggle button when hovered  */
                                .skin-blue .main-header .navbar .sidebar-toggle:hover{
                                background-color: #3C0753;
                                }

                                /* body */
                                .content-wrapper, .right-side {
                                background-color: #FFFFFF	;
                                }

                                '))),
      tabItems(
        tabItem(
          tabName = "home",
          fluidRow(
            box(title = "BLURB",
                status = "warning",
                solidHeader = TRUE,
                collapsible = TRUE,
                width = 6,
                p(style="text-align: justify;", "RareInsight is a project aimed at tackling the challenges posed by rare diseases through collaborative efforts in research. Developed as an open-source, interactive dashboard tailored for both clinicians and patients, RareInsight focuses on transforming the interpretation of genetic variant data.The project's core objective is to simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. RareInsight processes genetic variant data into customizable, interactive reports, ideally generated by nf-coreâ€™s raredisease pipeline. These reports are derived from whole genome or whole exome sequencing data and offer advanced filtering options, statistical analysis capabilities, and diverse export formats.Developed using Shiny and tested with data from respected sources such as the Undiagnosed Disease Program and NHGRI GREGOR Consortium datasets found in dbGaP, RareInsight ensures accuracy and reliability. RareInsight aids informed decision-making for clinicians and patients alike and fosters collaboration among researchers and clinicians, facilitating seamless sharing and collaboration on reports. By promoting knowledge exchange, RareInsight aims to enrich the collective understanding of rare diseases and enhance healthcare outcomes in an open-source format.In summary, RareInsight is poised to revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare."
                )),
            box(title = "Aim",
                status = "danger",
                solidHeader = TRUE,
                collapsible = TRUE,
                width = 6,
                h4("  By promoting knowledge exchange, RareInsight aims to : "),
                p("1- Simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. "),
                p("2- Processes genetic variant data into customizable, interactive reports."),
                p("3- Ensures accuracy and reliability.  "),
                p("4- Facilitating seamless sharing and collaboration on reports.  "),
                p("5- Enrich the collective understanding of rare diseases  "),
                p("6- Enhance healthcare outcomes in an open-source format. "),
                p("7- Revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare. ")
            )
          )
        ),

        tabItem(
          tabName = "service",
          h1("User Data Inputs")
        ),
        tabItem(
          tabName = "input_user_info",
          h1(" User Data Inputs"),
          fluidRow(
            column(6, textInput(inputId = "name", label = "Name")),
            column(6, textInput(inputId = "surname", label = "Surname")),
            column(6, textInput(inputId = "ethnicity", label = "Ethnicity")),
            column(6, dateInput(inputId = "dob", label = "Date of Birth")),
            column(12, textInput(inputId = "clinical_diagnosis", label = "Clinical Diagnosis (OMIM")),
            column(12, textInput(inputId = "phenotype", label = "Phenotype (HPO Terms)")),
            column(12, selectInput(inputId = "rest_performed", label = "Test Performed", choices = c("WES", "WGS", "Gene Panel"))),
            downloadButton(outputId = "download_button", label = "Download User Info")
          )
        ),

        tabItem(
          tabName = "Search",
          h1("Search Panel"),
          fluidRow(
            wellPanel(
              selectInput("search_type", "Search Type:",
                          choices = c("Gene", "Variant", "Disorder"),
                          selected = "Gene"),
              textInput("search_input", "Enter Search Term:", ""),
              actionButton("search_button", "Search"),
              DTOutput("search_results")
            )
          )
        ),

        tabItem(
          tabName = "vcfpanel",
          h1("VCF Panel"),
          # Added file input
          fileInput("input_file", "Upload VCF File", accept = c(".vcf.gz")),
          tabsetPanel(
            tabPanel("Results",
                     fluidRow(
                       wellPanel(
                         title = "Variant Information",
                         solidHeader = TRUE,
                         status = "primary",
                         div(style = "overflow-x: scroll; width: 100%;",
                             tableOutput("variant_table")
                         )
                       )
                     )
            ),
            tabPanel(
              "Graphs",
              fluidRow(

              )
            )
          )
        ),

        tabItem(
          tabName = "diagnostic_report",
          tabsetPanel(
            tabPanel(
              "Disorder Information",
              h2("Disorder Information"),
              p("This is the content for Disorder Information."),
              actionButton("buttonOMIM", label = "OMIM", icon = icon("up-right-from-square")),
              actionButton("buttonCLINVAR", label = "ClinVar", icon = icon("up-right-from-square")),
              actionButton("buttonPubMed", label = "PubMed", icon = icon("up-right-from-square")),
              actionButton("buttongenomAD", label = "genomAD", icon = icon("up-right-from-square")),
              actionButton("buttonSupportGroup", label = "Support group", icon = icon("up-right-from-square"))
            ),
            tabPanel(
              "Clinician and Researcher Support",
              h2("Clinician and Researcher Support"),
              p("This is the content for Clinician and Researcher Support.")
            ),
            tabPanel(
              "Patient Support",
              h2("Patient Support"),
              p("This is the content for Patient Support.")
            )
          )
        ),

        tabItem(
          tabName = "acknow",
          h1("Contributors:"),
          h4(HTML("<a href='https://github.com/Kimmiecc19'>KC Coetzer</a>: Division of Molecular Biology and Human Genetics, Department of Biomedical Sciences, Faculty of Medicine and Health Sciences, Stellenbosch University, Cape Town, South Africa")),
          h4(HTML("<a href='https://github.com/Zemzemfiras1'>F Zemzem</a>: Laboratory of Cytogenetics, Molecular Genetics and Biology of Reproduction CHU Farhat Hached Sousse, Higher Institute of Biotechnology of Monastir, University of Monastir, Tunisia")),
          h4(HTML("<a href='https://github.com/ssyamoako'>EK Amoako</a>: Genomics and Bioinformatics Laboratory, West African Centre for Cell Biology of Infectious Pathogens (WACCBIP), University of Ghana, Legon, Ghana")),
          h4(HTML("<a href='https://github.com/AkurutEva'>E Akurut</a>: African Centre of Excellence in Bioinformatics and Data-intensive Sciences, Infectious Disease Institute, Makerere University, Kampala, Uganda")),
          h4(HTML("<a href='https://github.com/gwiafe'>G Wiafe</a>: Department of Biomedical Sciences, University of Cape Coast, Cape Coast, Ghana")),
          h4(HTML("<a href='https://github.com/laitanawe'>OI Awe</a>: African Society for Bioinformatics and Computational Biology, Cape Town, South Africa")),
          h1("Thank you to the following organizations:"),
          h4("African Society for Bioinformatics and Computational Biology",
             img(src = "asbcb-logo.png", height = 50, width = 50))
        )
      )
    )
  )
)

# Define server logic
server <- function(input, output, session) {
  # Define variant_summary outside observe block
  variant_summary <- reactiveVal(NULL)

  observe({
    withProgress(
      message = 'Downloading and processing data...',
      detail = 'This may take a while...',
      value = 0, {

        # Download and process data
        download_file <- function(url, dest_file) {
          if (!file.exists(dest_file)) {
            tryCatch(
              {
                system(paste("curl -o", dest_file, url))
                cat("...done...\n")
              },
              error = function(e) {
                cat("Error:", e$message, "\n")
              }
            )
          } else {
            cat("...File already exists...\n")
          }
        }

        # Check if the file already exists
        if (!file.exists("variant_summary.txt.gz")) {
          # Download the latest ClinVar summary from FTP server
          cat("...downloading latest ClinVar summary from: https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz...\n")
          download_file("https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz", "variant_summary.txt.gz")
        } else {
          # File exists, so inform the user
          cat("...previous version found... skipping download....\n")
        }

        # Unzip the file
        cat("...unzipping the file...\n")
        system("gunzip -f variant_summary.txt.gz")  # Use -f option to force overwrite if the file already exists

        # Read the file
        variant_summary_data <- read.delim("variant_summary.txt", stringsAsFactors = FALSE)
        variant_summary(variant_summary_data)
      })
  })

  ## user input part :
  user_info_text <- reactive({
    paste(
      "Name: ", input$name,"\n",
      "Surname: ", input$surname,"\n",
      "Ethnicity: ", input$ethnicity,"\n",
      "Date of Birth: ", input$dob,"\n",
      "Clinical Diagnosis (OMIM): ", input$clinical_diagnosis,"\n",
      "Phenotype (HPO Terms): ", input$phenotype,"\n",
      "Test Performed: ", input$rest_performed,"\n"
    )
  })

  # Function to generate downloadable user info content
  output$download_button <- downloadHandler(
    filename = function() {
      paste("user_info_", Sys.Date(), ".txt", sep = "")
    },
    content = function(file) {
      writeLines(user_info_text(), file)
    }
  )

  ### Search Panel part
  observeEvent(input$search_button, {
    search_term <- input$search_input
    search_type <- input$search_type
    # Define the column to search based on search type
    search_column <- switch(
      search_type,
      "Gene" = "GeneSymbol",
      "Variant" = "Name",
      "Disorder" = "PhenotypeList"
    )
    # Filter data based on search term and column
    filtered_data <- variant_summary()[grep(search_term, variant_summary()[[search_column]], ignore.case = TRUE), ]
    output$search_results <- renderDT({
      if(nrow(filtered_data) == 0) {
        # Display a message if no results are found
        HTML("<p>No matching results found.</p>")
      } else {
        datatable(filtered_data, options = list(scrollX = TRUE))
      }
    })
  })
  # Add this inside the server function, after the UI and before the shinyApp() function
  
  observeEvent(input$buttonOMIM, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the OMIM search URL with the search term and open it in the browser
      omim_search_url <- paste0("https://www.omim.org/search?index=entry&start=1&search=", URLencode(search_term), "&sort=score+desc%2C+prefix_sort+desc&limit=10&date_created_from=&date_created_to=&date_updated_from=&date_updated_to=")
      browseURL(omim_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  ## VCF upload file part
  vcf_data <- reactive({
    req(input$input_file)
    vcf_file <- input$input_file$datapath
    vcf_data <- read.vcfR(vcf_file)
    return(vcf_data)
  })

  output$variant_table <- renderTable({
    req(vcf_data())

    # Extract variant information and filter PASS variants
    variant_info <- as.data.frame(vcf_data()@fix)
    ############ variant_info <- variant_info[variant_info$FILTER == "PASS", ]

    # Remove ID and QUAL columns
    ############ variant_info <- subset(variant_info, select = -c(ID, QUAL))

    # Remobe ID is not Abvailable "NA"
    if (any(is.na(variant_info$ID))) {
      variant_info <- subset(variant_info, select = -c(ID))
    }

    # Split INFO column into separate columns
    info_split <- strsplit(as.character(variant_info$INFO), ";")
    max_splits <- max(lengths(info_split))

    # Create dataframe to store split columns
    info_df <- data.frame(matrix(NA, ncol = max_splits, nrow = nrow(variant_info)))

    # Extract column names and values from INFO column
    info_names_values <- lapply(info_split, function(x) {
      splits <- strsplit(x, "=")
      names <- sapply(splits, `[`, 1)
      values <- sapply(splits, `[`, 2)
      return(list(names = names, values = values))
    })

    # Fill in split columns with appropriate column names
    for (i in 1:max_splits) {
      column_values <- sapply(info_names_values, function(info) ifelse(length(info$values) >= i, info$values[i], NA))
      col_names <- unique(unlist(sapply(info_names_values, function(info) ifelse(length(info$names) >= i, info$names[i], NA))))
      colnames(info_df)[i] <- col_names
      info_df[, i] <- column_values
    }

    # Combine split INFO columns with original dataframe
    variant_info <- cbind(variant_info, info_df)

    # Remove original INFO column
    variant_info <- subset(variant_info, select = -INFO)

    return(variant_info)
  })
}

shinyApp(ui, server)
