library(shiny)
library(shinydashboard)
library(vcfR)
library(DT)  # For interactive tables

ui <- dashboardPage(
  dashboardHeader(
    title = span(img(src = "3.png", height = 50), "RareInsight")
  ),
  dashboardSidebar(
    sidebarMenu(
      menuItem("Home", tabName = "home", icon = icon("home")),
      menuItem("About us", tabName = "Abt_us", icon = icon("users")),
      menuItem("Services", tabName = "service", icon = icon("bar-chart"),
               menuSubItem("Input User Information", tabName = "input_user_info"),
               menuSubItem("Search Panel", tabName = "search"),
               menuSubItem("VCF analysis", tabName = "vcfpanel"),
               menuSubItem("Diagnostic report", tabName = "diagnostic_report")
      ),
      menuItem("Acknowledgement", tabName = "acknow", icon = icon("handshake"))
    )
  ),
  dashboardBody(
    tags$head(tags$style(HTML('
                                /* logo */
                                .skin-blue .main-header .logo {
                                background-color: #030637;
                                }

                                /* navbar (rest of the header) */
                                .skin-blue .main-header .navbar {
                                background-color: #030637;
                                }

                                /* logo when hovered */
                                .skin-blue .main-header .logo:hover {
                                background-color: #030637;
                                }



                                /* main sidebar */
                                .skin-blue .main-sidebar {
                                background-color: #3C0753;
                                }

                                /* active selected tab in the sidebarmenu */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu .active a{
                                background-color: #9f16db;
                                }

                                /* other links in the sidebarmenu */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu a{
                                background-color: ;
                                color: #FFFFFF
                                }

                                /* other links in the sidebarmenu when hovered */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu a:hover{
                                background-color: #030637;
                                }
                                /* toggle button when hovered  */
                                .skin-blue .main-header .navbar .sidebar-toggle:hover{
                                background-color: #3C0753;
                                }

                                /* body */
                                .content-wrapper, .right-side {
                                background-color: #FFFFFF	;
                                }

                                '))),

    tabItems(
      tabItem(
        tabName = "home",
        fluidRow(
          box(title = "Summary",
              status = "warning",
              solidHeader = TRUE,
              collapsible = TRUE,
              width = 6,
              p(style="text-align: justify;", "RareInsight is a project aimed at tackling the challenges posed by rare diseases through collaborative efforts in research. Developed as an open-source, interactive dashboard tailored for both clinicians and patients, RareInsight focuses on transforming the interpretation of genetic variant data.The project's core objective is to simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. RareInsight processes genetic variant data into customizable, interactive reports, ideally generated by nf-coreâ€™s raredisease pipeline. These reports are derived from whole genome or whole exome sequencing data and offer advanced filtering options, statistical analysis capabilities, and diverse export formats.Developed using Shiny and tested with data from respected sources such as the Undiagnosed Disease Program and NHGRI GREGOR Consortium datasets found in dbGaP, RareInsight ensures accuracy and reliability. RareInsight aids informed decision-making for clinicians and patients alike and fosters collaboration among researchers and clinicians, facilitating seamless sharing and collaboration on reports. By promoting knowledge exchange, RareInsight aims to enrich the collective understanding of rare diseases and enhance healthcare outcomes in an open-source format.In summary, RareInsight is poised to revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare."
              )),
          box(title = "Aim",
              status = "danger",
              solidHeader = TRUE,
              collapsible = TRUE,
              width = 6,
              h4("  By promoting knowledge exchange, RareInsight aims to : "),
              p("1- Simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. "),
              p("2- Processes genetic variant data into customizable, interactive reports."),
              p("3- Ensures accuracy and reliability.  "),
              p("4- Facilitating seamless sharing and collaboration on reports.  "),
              p("5- Enrich the collective understanding of rare diseases  "),
              p("6- Enhance healthcare outcomes in an open-source format. "),
              p("7- Revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare. ")
          )
        )
      ),
      tabItem(
        tabName = "Abt_us",
        h1("Introduction for members"),
        h3("It should be the last tab")
      ),
      tabItem(
        tabName = "service",
        h1("Services")
      ),
      tabItem(
        tabName = "diagnostic_report",
        tabPanel("Clinician and Researcher Support", "Clinician and researcher support information will be displayed here"),
        tabPanel("Patient Support", "Patient support information will be displayed here"),
        tabPanel("ACMG Classification", "ACMG classification information will be displayed here"),
        tabPanel("Disorder", "Clinical information will be displayed here")
      ),
      tabItem(
        tabName = "input_user_info",
        h1("User Data Inputs"),
        fluidRow(
          column(6, textInput("name", "Name")),
          column(6, textInput("surname", "Surname")),
          column(6, textInput("ethnicity", "Ethnicity")),
          column(6, textInput("doctor", "Doctor")),
          column(6, dateInput("dob", "Date of Birth", value = Sys.Date())),
          column(12, textInput("clinical_diagnosis", "Clinical Diagnosis (OMIM)")),
          column(12, textInput("phenotype", "Phenotype (HPO Terms)")),
          column(12, selectInput("test_performed", "Test Performed", choices = c("WES", "WGS", "Gene Panel"))),
          column(12, actionButton("download_button", "Download User Info")),
          column(12, uiOutput("download_markdown"))
        )
      ),
      tabItem(
        tabName = "vcfpanel",
        h1("VCF Panel"),
        fluidPage(
          fileInput("input_file", "Upload VCF File", accept = c(".vcf.gz")),
          fluidRow(
            column(
              width = 12,
              box(
                title = "Variant Information",
                solidHeader = TRUE,
                status = "primary",
                div(
                  style = "overflow-x: scroll; width: 100%;",
                  tableOutput("variant_table")
                )
              )
            )
          )
        )
      ),
      tabItem(
        tabName = "acknow",
        h1("Thank you to the following people and organizations:"),
        h4("African Society for Bioinformatics and Computational Biology",
           img(src = "asbcb-logo.png", height = 50, width = 50)
        )
      )
    )
  )
)

variant_summary <- read.delim("variant_summary.txt", stringsAsFactors = FALSE)

server <- function(input, output, session) {
  observeEvent(input$search_button, {
    search_term <- input$search_input
    search_type <- input$search_type
    # Define the column to search based on search type
    search_column <- switch(
      search_type,
      "Gene" = "GeneSymbol",
      "Variant" = "Name",
      "Disorder" = "PhenotypeList"
    )
    # Filter data based on search term and column
    filtered_data <- variant_summary[grep(search_term, variant_summary[[search_column]], ignore.case = TRUE), ]
    output$search_results <- renderDT({
      if(nrow(filtered_data) == 0) {
        # Display a message if no results are found
        HTML("<p>No matching results found.</p>")
      } else {
        datatable(filtered_data, options = list(scrollX = TRUE))
      }
    })
  })

  user_info <- reactive({
    data.frame(
      Name = input$name,
      Surname = input$surname,
      Ethnicity = input$ethnicity,
      Doctor = input$doctor,
      DateOfBirth = input$dob,
      ClinicalDiagnosis = input$clinical_diagnosis,
      Phenotype = input$phenotype,
      TestPerformed = input$test_performed
    )
  })

  output$download_markdown <- renderUI({
    downloadButton("download_button", "Download Markdown")
  })

  observeEvent(input$download_button, {
    output_file <- paste("user_info_", Sys.Date(), ".md", sep = "")
    writeLines(c("# User Information\n\n```{r}", paste(names(user_info()), collapse=", "), "\n", unlist(user_info()), "\n```"), output_file)
  })

  vcf_data <- reactive({
    req(input$input_file)
    vcf_file <- input$input_file$datapath
    vcf_data <- read.vcfR(vcf_file)
    return(vcf_data)
  })

  output$variant_table <- renderTable({
    req(vcf_data())

    # Extract variant information and filter PASS variants
    variant_info <- as.data.frame(vcf_data()@fix)
    variant_info <- variant_info[variant_info$FILTER == "PASS", ]

    # Remove ID and QUAL columns
    variant_info <- subset(variant_info, select = -c(ID, QUAL))

    # Split INFO column into separate columns
    info_split <- strsplit(as.character(variant_info$INFO), ";")
    max_splits <- max(lengths(info_split))

    # Create dataframe to store split columns
    info_df <- data.frame(matrix(NA, ncol = max_splits, nrow = nrow(variant_info)))

    # Extract column names and values from INFO column
    info_names_values <- lapply(info_split, function(x) {
      splits <- strsplit(x, "=")
      names <- sapply(splits, `[`, 1)
      values <- sapply(splits, `[`, 2)
      return(list(names = names, values = values))
    })

    # Fill in split columns with appropriate column names
    for (i in 1:max_splits) {
      column_values <- sapply(info_names_values, function(info) ifelse(length(info$values) >= i, info$values[i], NA))
      col_names <- unique(unlist(sapply(info_names_values, function(info) ifelse(length(info$names) >= i, info$names[i], NA))))
      colnames(info_df)[i] <- col_names
      info_df[, i] <- column_values
    }

    # Combine split INFO columns with original dataframe
    variant_info <- cbind(variant_info, info_df)

    # Remove original INFO column
    variant_info <- subset(variant_info, select = -INFO)

    return(variant_info)
  })
}

shinyApp(ui, server)
