library(shiny)
library(shinydashboard)
library(vcfR)

ui <- fluidPage(
  dashboardPage(
    skin = "red",

    dashboardHeader(title = "RareInsight"),

    dashboardSidebar(
      sidebarMenu(
        menuItem("Home ", tabName = "home", icon = icon("home")),
        menuItem("About us", tabName = "Abt_us", icon = icon("users")),
        menuItem("Services", tabName = "service", icon = icon("bar-chart"),
                 menuSubItem("Search Panel",
                             tabName = "Search",
                             icon = icon("glyphicon glyphicon-stats" )),
                 menuSubItem("Vcf Panel",tabName = "vcfpanel",icon = icon("file")) 
        ),
      menuItem("Acknowledgement", tabName = "acknow", icon = icon("handshake"))
      )
    ),

    dashboardBody(
      tags$style(
        HTML(
          "
        /* CSS to add a background image to the dashboard body */
        .content-wrapper {
        background-image: url('https://pbs.twimg.com/media/EvD_E0uWYAI_sFn.jpg'); /* Replace 'your_image_url.jpg' with your image URL */
        background-size: cover;
        background-repeat: no-repeat;
        background-position: center;
        }
        "
        )
      ),

      tabItems(
        tabItem(
          tabName = "home",
          fluidRow(
            box(title = "BLURB",
                status = "warning",
                solidHeader = TRUE,
                collapsible = TRUE,
                width = 6,
                p("RareInsight is a project aimed at tackling the challenges posed by rare diseases through collaborative efforts in research. Developed as an open-source, interactive dashboard tailored for both clinicians and patients, RareInsight focuses on transforming the interpretation of genetic variant data.The project's core objective is to simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. RareInsight processes genetic variant data into customizable, interactive reports, ideally generated by nf-coreâ€™s raredisease pipeline. These reports are derived from whole genome or whole exome sequencing data and offer advanced filtering options, statistical analysis capabilities, and diverse export formats.Developed using Shiny and tested with data from respected sources such as the Undiagnosed Disease Program and NHGRI GREGOR Consortium datasets found in dbGaP, RareInsight ensures accuracy and reliability. RareInsight aids informed decision-making for clinicians and patients alike and fosters collaboration among researchers and clinicians, facilitating seamless sharing and collaboration on reports. By promoting knowledge exchange, RareInsight aims to enrich the collective understanding of rare diseases and enhance healthcare outcomes in an open-source format.In summary, RareInsight is poised to revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare."
                )),
            box(title = "Aim",
                status = "danger",
                solidHeader = TRUE,
                collapsible = TRUE,
                width = 6,
                h4("  By promoting knowledge exchange, RareInsight aims to : "),
                p("1- Simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. "),
                p("2- Processes genetic variant data into customizable, interactive reports."),
                p("3- Ensures accuracy and reliability.  "),
                p("4- Facilitating seamless sharing and collaboration on reports.  "),
                p("5- Enrich the collective understanding of rare diseases  "),
                p("6- Enhance healthcare outcomes in an open-source format. "),
                p("7- Revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare. ")
            )
          )
        ),

        tabItem(
          tabName = "Abt_us",
          h1("Intoduction for members "),
          h3("it should be the last tab")
        ),

        tabItem(
          tabName = "service",
          h1("User Data Inputs")
        ),

        tabItem(
          tabName = "Search",
          h1("Search Panel"),
          tabBox(
            width = 6,
            textInput("disorder_name", "Enter genetic disorder name"),
            textInput("hgvs_variant", "Enter HGVS variant"),
            actionButton("proceed_button", "Proceed")
          ),
          tabBox(
            title = "Results",side = c("right"),
            id = "tabset1",
            width = 12,
            tabPanel("Export", "Export options will be displayed here"),
            tabPanel("Clinician and Researcher Support", "Clinician and researcher support information will be displayed here"),
            tabPanel("Patient Support", "Patient support information will be displayed here"),
            tabPanel("ACMG Classification", "ACMG classification information will be displayed here"),
            tabPanel("Disorder", "Clinical information will be displayed here")
          )
        ),
        tabItem(tabName = "vcfpanel",
                h1("VCF Panel"),
                
                # Added file input
                fileInput("input_file", "Upload VCF File", accept = c(".vcf.gz")),
                fluidRow(
                  box(
                    title = "Variant Information",
                    solidHeader = TRUE,
                    status = "primary",
                    div(style = "overflow-x: scroll; width: 100%;",
                        tableOutput("variant_table")
                    )
                  )
                )
          
        ),

        tabItem(
          tabName = "acknow",
          h1("Acknowledgement part for omics codeathon universities & contribution")
        )
      )

      
    )
  )
)

server <- function(input, output, session) {
  vcf_data <- reactive({
    req(input$input_file)
    vcf_file <- input$input_file$datapath
    vcf_data <- read.vcfR(vcf_file)
    return(vcf_data)
  })

  output$variant_table <- renderTable({
    req(vcf_data())
    variant_info <- as.data.frame(vcf_data()@fix)
    return(variant_info)
  })
}

shinyApp(ui, server)
