#Check if packages are downloaded
if (!require("shiny")) install.packages("shiny")
if (!require("shinydashboard")) install.packages("shinydashboard")
if (!require("progress")) install.packages("progress")
if (!require("data.table")) install.packages("data.table")
if (!require("vcfR")) install.packages("vcfR")
if (!require("DT")) install.packages("DT")
if (!require("rmarkdown")) install.packages("rmarkdown")
if (!require("shinyjs")) install.packages("shinyjs")

#Load packages
library(shiny)
library(shinydashboard)
library(progress)
library(data.table)
library(vcfR)
library(DT)  # For interactive tables
library(rmarkdown)
library(shinyjs)


#Set UI function
ui <- fluidPage(
  dashboardPage(
    dashboardHeader(title = span(img(src = "rareinsight_final.png", height = 50), "RareInsight")),
    dashboardSidebar(
      sidebarMenu(
        menuItem("Home ", tabName = "home", icon = icon("home")),
        menuItem("Understanding your genetics", tabName = "understanding_genetics", icon = icon("education", lib = "glyphicon")),
        menuItem("Services", tabName = "service", icon = icon("bar-chart"),
                 menuSubItem("Input User Information", tabName = "input_user_info", icon = icon("user-circle")),
                 menuSubItem("Vcf Panel", tabName = "vcfpanel", icon = icon("file")),
                 menuSubItem("Search Panel", tabName = "Search", icon = icon("search", lib = "glyphicon" )),
                 menuItem("Diagnostic report", tabName = "diagnostic_report", icon = icon("line-chart"))
        ),
        menuItem("Acknowledgement", tabName = "acknow", icon = icon("handshake"))
      )
    ),
    dashboardBody(
      tags$head(tags$style(HTML('
                                /* logo */
                                .skin-blue .main-header .logo {
                                background-color: #030637;
                                }

                                /* navbar (rest of the header) */
                                .skin-blue .main-header .navbar {
                                background-color: #030637;
                                }

                                /* logo when hovered */
                                .skin-blue .main-header .logo:hover {
                                background-color: #030637;
                                }

                                /* main sidebar */
                                .skin-blue .main-sidebar {
                                background-color: #3C0753;
                                }

                                /* active selected tab in the sidebarmenu */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu .active a{
                                background-color: #9f16db;
                                }

                                /* other links in the sidebarmenu */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu a{
                                background-color: ;
                                color: #FFFFFF
                                }

                                /* other links in the sidebarmenu when hovered */
                                .skin-blue .main-sidebar .sidebar .sidebar-menu a:hover{
                                background-color: #030637;
                                }
                                /* toggle button when hovered  */
                                .skin-blue .main-header .navbar .sidebar-toggle:hover{
                                background-color: #3C0753;
                                }

                                /* body */
                                .content-wrapper, .right-side {
                                background-color: #FFFFFF	;
                                }

                                '))),
      tabItems(
        tabItem(
          tabName = "home",
          fluidRow(
            box(title = "BLURB",
                status = "warning",
                solidHeader = TRUE,
                collapsible = TRUE,
                width = 6,
                p(style="text-align: justify;", "RareInsight is a project aimed at tackling the challenges posed by rare diseases through collaborative efforts in research. Developed as an open-source, interactive dashboard tailored for both clinicians and patients, RareInsight focuses on transforming the interpretation of genetic variant data.The project's core objective is to simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. RareInsight processes genetic variant data into customizable, interactive reports, ideally generated by nf-core’s raredisease pipeline. These reports are derived from whole genome or whole exome sequencing data and offer advanced filtering options, statistical analysis capabilities, and diverse export formats.Developed using Shiny and tested with data from respected sources such as the Undiagnosed Disease Program and NHGRI GREGOR Consortium datasets found in dbGaP, RareInsight ensures accuracy and reliability. RareInsight aids informed decision-making for clinicians and patients alike and fosters collaboration among researchers and clinicians, facilitating seamless sharing and collaboration on reports. By promoting knowledge exchange, RareInsight aims to enrich the collective understanding of rare diseases and enhance healthcare outcomes in an open-source format.In summary, RareInsight is poised to revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare."
                )),
            box(title = "Aim",
                status = "danger",
                solidHeader = TRUE,
                collapsible = TRUE,
                width = 6,
                h4("  By promoting knowledge exchange, RareInsight aims to : "),
                p("1- Simplify the complex process of diagnosing rare diseases and improving therapeutic outcomes. "),
                p("2- Processes genetic variant data into customizable, interactive reports."),
                p("3- Ensures accuracy and reliability.  "),
                p("4- Facilitating seamless sharing and collaboration on reports.  "),
                p("5- Enrich the collective understanding of rare diseases  "),
                p("6- Enhance healthcare outcomes in an open-source format. "),
                p("7- Revolutionize rare disease diagnosis and research by leveraging collaboration and innovation, ultimately reshaping the landscape of healthcare. ")
            )
          )
        ),
        
        tabItem(
          tabName = "understanding_genetics",
          tabsetPanel(
            tabPanel(
              "Glossary",
              h2("Glossary"),
              p(HTML(style="text-align: justify;", "<a href='https://www.genomicseducation.hee.nhs.uk/glossary/'>Here</a> are some important terms to start with.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Allele:</b></h4>Variants of a gene found at the same position on a chromosome. They can be identical or different, and each variant creates a new allele. Some alleles have functional effects, like the numerous pathogenic variants in the BRCA1 gene.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Cell:</b></h4>The fundamental unit of life, enclosed by a membrane and containing organelles such as the nucleus. The cell's structure and function are determined by its DNA and environment, with changes in DNA potentially affecting both.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Chromosome:</b></h4>Bundles of DNA and associated proteins, with humans typically having 46 chromosomes in 23 pairs. Chromosomes contain nearly all of a cell's DNA, and changes to them can result in various conditions, including trisomies and structural abnormalities.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Diagnostic genomic test:</b></h4>A test providing a diagnosis for a condition caused by genomic changes. It confirms or refutes a diagnosis, with various technologies available, including targeted tests and broader approaches like whole exome or genome sequencing.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Diagnostic odyssey:</b></h4>A prolonged journey to receive a diagnosis for a medical condition, especially rare ones. It involves multiple tests, sometimes invasive, and can take years, often ending with genomic testing providing a conclusive diagnosis.")),
              h4(HTML(style="text-align: justify;", "<h4><b>DNA:</b></h4>The chemical carrying genetic information, made up of four bases: adenine, cytosine, guanine, and thymine. It contains all the instructions necessary for an organism's growth and maintenance, with changes in DNA having significant effects on an organism's characteristics.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Double helix:</b></h4>The structure formed by two DNA strands, with nucleotides arranged in a helical pattern. This structure aids in replication, with most DNA naturally occurring as a right-handed helix.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Exome:</b></h4>The portion of the genome consisting of exons, which encode proteins. It represents less than 2% of the human genome but is well understood compared to other regions.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Gene:</b></h4>A segment of DNA encoding a specific polypeptide, typically a protein or part of a protein. Genes play crucial roles in cellular function, and alterations within genes can lead to dysfunctional or absent protein products.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Genetics:</b></h4>The study of genes and their inheritance, often used interchangeably with genomics. Genetics focuses on individual genes, while genomics encompasses all DNA needed for an organism's development and function.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Genome:</b></h4>An organism's complete genetic material, encompassing both genes and non-coding sequences. Sequencing genomes aids in healthcare applications like disease prediction, diagnosis, and treatment.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Mutation:</b></h4>Any difference in a gene's sequence compared to a reference genome, potentially leading to altered gene function and disease. It can be synonymous with the term variant, particularly when referring to pathogenic alterations.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Next-generation sequencing:</b></h4>Modern high-throughput sequencing techniques reading millions of DNA or RNA fragments simultaneously. It allows for faster, cost-efficient sequencing and has broadened applications in routine healthcare.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Phenotype:</b></h4>An organism's observable characteristics, influenced by genetics and environment. It includes both physical traits and symptoms of conditions, guiding clinical investigations and diagnostics.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Sequencing:</b></h4>Laboratory technique determining the order of bases in DNA, crucial for various clinical contexts, with methods including Sanger sequencing and next-generation sequencing.")),
              h4(HTML(style="text-align: justify;", "<h4><b>Variant:</b></h4>Any difference between two individuals' genomes or a reference genome, with variants having varying impacts, from harmless to disease-causing. They can range from single base changes to larger alterations."))
            ),
            tabPanel(
              "Basic Genetics",
              div(
                h2(style="text-align: justify;", "What is Genetics?"),
                h4(style="text-align: justify;", "Genomics = DNA = You"),
                h4(style="text-align: justify;", "Genetics is the study of how traits are passed from parents to children. It involves understanding DNA, which contains the genetic instructions for building and maintaining our bodies."),
                h4(style="text-align: justify;", "Genomics on the other hand is the study of an organism’s genome – its genetic material – and how that information is applied. All living things (plants, animals, humans etc.) have a genome – and ours is made up of DNA."),
                h4(img(src = "https://www.genomicseducation.hee.nhs.uk/wp-content/uploads/2019/06/Half-plus-half-equals-you-960x640.png", height = 300, width = 300))
              ),
              div(
                h2(style="text-align: justify;", "What makes up our genome?"),
                h4(style="text-align: justify;", "Your genome is unique. A copy is found in almost every cell in your body and is organised into 46 chromosomes in 23 pairs. These chromosomes are made up of genes"),
                h4(style="text-align: justify;", "Some sections of our DNA are called genes. Humans have around 20,000 'coding' genes, which contain the information to build proteins. These proteins are essential for building and repairing our bodies"),
                h4(style="text-align: justify;", "Some genes determine physical characteristics, such as eye colour or height. This is also called our phenotype. Other genes can influence the chance of developing a health condition, such as cystic fibrosis or skeletal dysplasia."),
                h4(img(src = "https://rareshareprod.s3.amazonaws.com/uploads/ckeditor/pictures/51/content_Chromosomes.png", height = 300, width = 300)),
                h4(img(src = "https://www.genomicseducation.hee.nhs.uk/wp-content/uploads/2018/08/DNA-double-helix_13081113544_l.jpg", height = 300, width = 300))
              ),
              div(
                h2(style="text-align: justify;", "How do we study our genome: sequencing"),
                h4(style="text-align: justify;", "Reading the 3 billion letters in a human genome is a huge task, and the process generates a vast amount of data. Sequencing one person’s whole genome amounts to around 200GB of data – the capacity of a typical computer."),
                h4(style="text-align: justify;", "What happens to all that data? In the case of a genomic test, the data is filtered, analysed, interpreted, validated and assessed, alongside the patient’s health information, by a multidisciplinary team of health professionals, including bioinformaticians, clinical scientists and clinical geneticists. Only then can relevant results from a genomic test be passed on to the clinical team and their patient."),
                h4(style="text-align: justify;", "As the genome is passed from both parent to offspring and from cell to cell in our body, any change in the DNA - known as a variant - can also be passed on. In healthcare, this becomes significant if the variant is associated with a particular medical condition which can result in a rare disease.")
              ),
              div(
                h2(style="text-align: justify;", "How do we get these variants?"),
                h4(style="text-align: justify;", "Genes are passed down from our parents"),
                h4(style="text-align: justify;", "Every person has 2 copies of each gene with 23 chromosomes from your mother and 23 chromosomes from your father in each gene. Approximately 99.9% of the genes in your genome are identical from person to person. The remaining genes are known as variants. These variants can be seen as “spelling mistakes” in your genome and can either be very small (one letter) or very large (a whole sentence). Sometimes, these spelling mistakes can cause a genetic disorder and some of these disorders are very rare."),
                h4(img(src = "https://cdn.bluebirdbio.com/-/media/Gene_Therapy/com/1/understanding-genetic-diseases/gene-inheritance-example.png?mw=522&rev=ac214731a4824980a5f5a0cf8df4b8a0&hash=DAE97500D51B477B45304E0B24BD7357", height = 300, width = 300))
              )
            ),
            tabPanel(
              "Rare Diseases",
              div(
                h2(style="text-align: justify;", "What are Rare Diseases?"),
                h4(style="text-align: justify;", "A rare disease is a condition that affects fewer than 1 in 2,000 people."),
                h4(style="text-align: justify;", "More than 7,000 rare diseases have been recorded in over 7000 different genes."),
                h4(style="text-align: justify;", "Around 80% of rare diseases are caused by your genetics with around 70% manifesting in childhood."),
                h4(style="text-align: justify;", "While these disorders are classified as rare, they can collectively affect 6–8% of the general population."),
                h4(style="text-align: justify;", "Most people with a rare disease will face what is known as a ‘diagnostic odyssey’: the long journey a patient goes through to reach a diagnosis. These patients often see numerous healthcare professionals. Even if an accurate diagnosis is reached, the journey usually doesn’t end for the patient. This is because there is often a lack of available treatment options, a limited awareness of a patient’s condition and insufficient support available for patients and their families.")
              ),
              div(
                h2(style="text-align: justify;", "How are these variants inherited?"),
                h4(style="text-align: justify;", "There are three main types of changes that can cause a genetic disorder:"),
                h4(style="text-align: justify;", "Single gene conditions: caused by changes to one gene."),
                h4(style="text-align: justify;", "Chromosome conditions result from changes in the number or structure of the chromosomes."),
                h4(style="text-align: justify;", "Multifactorial conditions (or complex diseases) are caused by changes in multiple genes, often in a complex interaction with environmental factors."),
                h4(style="text-align: justify;", "These changes are inherited in 3 main ways"),
                h4(img(src = "https://rareshareprod.s3.amazonaws.com/uploads/ckeditor/pictures/50/content_DomRessive.png", height = 300, width = 300)),
                h4(style="text-align: justify;", "Dominant conditions happen when a person has one “healthy” copy and one mutated copy of a gene. If one parent has the mutation, there is a 50% chance of passing it on to a child in each pregnancy. Sometimes a patient has a mutated gene but neither parent has the same mutation. This is known as a de novo variant."),
                h4(style="text-align: justify;", "Recessive conditions only occur when an individual has two mutated copies of a gene. If a person has only one copy of this mutated gene, they are known as a carrier of the condition and may pass it to their children."),
                h4(style="text-align: justify;", "X-linked conditions are caused by genes mutated on your X chromosome. Females have two X chromosomes while males only have on X chromosome and one Y chromosome. Because of this, X-linked conditions more commonly affect males.")
              )
            )
          )
        ),
        
        tabItem(
          tabName = "service",
          h1("User Data Inputs")
        ),
        tabItem(
          tabName = "input_user_info",
          h1(" User Data Inputs"),
          fluidRow(
            column(6, textInput(inputId = "name", label = "Name")),
            column(6, textInput(inputId = "surname", label = "Surname")),
            column(6, textInput(inputId = "ethnicity", label = "Ethnicity")),
            column(6, dateInput(inputId = "dob", label = "Date of Birth")),  #TODO: User must be able to select year from search 
            column(12, textInput(inputId = "clinical_diagnosis", label = "Clinical Diagnosis")),
            column(12, textInput(inputId = "phenotype", label = "Phenotype")),
            column(12, selectInput(inputId = "test_performed", label = "Test Performed", choices = c("WES", "WGS", "Gene Panel"))),
            downloadButton(outputId = "download_button", label = "Download User Info")
          )
        ),

        tabItem(
          tabName = "Search",
          h1("Search Panel"),
          fluidRow(
            wellPanel(
              selectInput("search_type", "Search Type:",
                          choices = c("Gene", "Variant", "Disorder","dbSNP", "ClinVar"),
                          selected = "Gene"),
              textInput("search_input", "Enter Search Term:", ""),
              actionButton("search_button", "Search"),
              DTOutput("search_results")
            )
          )
        ),

        tabItem(
          tabName = "vcfpanel",
          h1("VCF Panel"),
          # Added file input
          fileInput("input_file", "Upload VCF File", accept = c(".vcf.gz")),
          tabsetPanel(
            tabPanel("Results",
                     fluidRow(
                       wellPanel(
                         title = "Variant Information",
                         solidHeader = TRUE,
                         status = "primary",
                         div(style = "overflow-x: scroll; width: 100%;",
                             tableOutput("variant_table")
                         )
                       )
                     )
            ),
            tabPanel(
              "Graphs",
              fluidRow(
                tabsetPanel(
                  box("plot1", plotOutput("PLOT1")),
                  box("plo2", plotOutput("PLOT2"))
                )
              )
            )
          )
        ),

        tabItem(
          tabName = "diagnostic_report",
          tabsetPanel(
            tabPanel(
              "Disorder Information",
              h2("Disorder Information"),
              p("THere is some information on the selected disorder, accession or gene from the Clinvar search panel."),
              ),
            tabPanel(
              "Clinician and Researcher Support",
              h2("Information for clinician and researchers"),
              p("This is the content for Clinician and Researcher Support."),
              h2("If you typed in a Gene name, you can search the following links:"),
              actionButton("buttonOMIM", label = "OMIM", icon = icon("up-right-from-square")),
              actionButton("buttonClinVar", label = "ClinVar", icon = icon("up-right-from-square")),
              actionButton("buttongenomAD", label = "genomAD", icon = icon("up-right-from-square")),
               actionButton("buttonPubMed", label = "PubMed", icon = icon("up-right-from-square")),
              actionButton("buttonGeneReviews", label = "GeneReviews", icon = icon("up-right-from-square")),
              h2("If you typed in a dbSNP ID or Clinvar accession in the search panel, you can search the following links:"),
              actionButton("buttonVarsome", label = "Varsome", icon = icon("up-right-from-square")),
              actionButton("buttonLitVar", label = "LitVar", icon = icon("up-right-from-square")),
              actionButton("buttondbSNP", label = "dbSNP", icon = icon("up-right-from-square")),
            #  actionButton("buttonClinVar", label = "ClinVar", icon = icon("up-right-from-square"))
            ),
            tabPanel(
              "Patient Support",
              h2("Patient Support"),
              p("Support page for rare disease patients and families"),
              actionButton("buttonRareConnect", label = "Rare Connect", icon = icon("up-right-from-square")),
              actionButton("buttonRAReSOURCE", label = "RAReSOURCE", icon = icon("up-right-from-square"))
            )
          )
        ),

        tabItem(
          tabName = "acknow",
          h2("Contributors:"),
          h4(HTML("<a href='https://github.com/Kimmiecc19'>KC Coetzer</a>: Division of Molecular Biology and Human Genetics, Department of Biomedical Sciences, Faculty of Medicine and Health Sciences, Stellenbosch University, Cape Town, South Africa")),
          h4(HTML("<a href='https://github.com/Zemzemfiras1'>F Zemzem</a>: Laboratory of Cytogenetics, Molecular Genetics and Biology of Reproduction CHU Farhat Hached Sousse, Higher Institute of Biotechnology of Monastir, University of Monastir, Tunisia")),
          h4(HTML("<a href='https://github.com/ssyamoako'>EK Amoako</a>: Genomics and Bioinformatics Laboratory, West African Centre for Cell Biology of Infectious Pathogens (WACCBIP), University of Ghana, Legon, Ghana")),
          h4(HTML("<a href='https://github.com/AkurutEva'>E Akurut</a>: African Centre of Excellence in Bioinformatics and Data-intensive Sciences, Infectious Disease Institute, Makerere University, Kampala, Uganda")),
          h4(HTML("<a href='https://github.com/gwiafe'>G Wiafe</a>: Department of Biomedical Sciences, University of Cape Coast, Cape Coast, Ghana")),
          h4(HTML("<a href='https://github.com/laitanawe'>OI Awe</a>: African Society for Bioinformatics and Computational Biology, Cape Town, South Africa")),
          h2("Thank you to the following organizations for their support:"),
          h4("African Society for Bioinformatics and Computational Biology", img(src = "asbcb-logo.png", height = 50, width = 50)),
          h2("For more information: "),
          h4(HTML("<a href='https://github.com/omicscodeathon/rareinsight/'>GitHub</a>: Read more about this research and access all our scripts."))
        )
      )
    )
  )
)

# Define server logic
server <- function(input, output, session) {
  # Define variant_summary outside observe block
  variant_summary <- reactiveVal(NULL)

  observe({
    withProgress(
      message = 'Downloading and processing data...',
      detail = 'This may take a while...',
      value = 0, {

        # Download and process data
        download_file <- function(url, dest_file) {
          if (!file.exists(dest_file)) {
            tryCatch(
              {
                system(paste("curl -o", dest_file, url))
                cat("...done...\n")
              },
              error = function(e) {
                cat("Error:", e$message, "\n")
              }
            )
          } else {
            cat("...File already exists...\n")
          }
        }

        # Check if the file already exists
        if (!file.exists("variant_summary.txt")) {
          # Download the latest ClinVar summary from FTP server
          cat("...downloading latest ClinVar summary from: https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz...\n")
          download_file("https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz", "variant_summary.txt.gz")
        } else {
          # File exists, so inform the user
          cat("...previous version found... skipping download....\n")
        }

        # Unzip the file
        cat("...unzipping the file...\n")
        system("gunzip -f variant_summary.txt.gz")  # Use -f option to force overwrite if the file already exists

        # Read the file
        variant_summary_data <- read.delim("variant_summary.txt", stringsAsFactors = FALSE)
        variant_summary(variant_summary_data)
      })
  })

  ## user input part : 
  user_info_text <- reactive({
    paste(
      "Name: ", input$name,"\n",
      "Surname: ", input$surname,"\n",
      "Ethnicity: ", input$ethnicity,"\n",
      "Date of Birth: ", input$dob,"\n",
      "Clinical Diagnosis (OMIM): ", input$clinical_diagnosis,"\n",
      "Phenotype (HPO Terms): ", input$phenotype,"\n",
      "Test Performed: ", input$test_performed,"\n"
    )
  })
  
  # Downloadable user info content
  output$download_button <- downloadHandler(
        filename = function() {
          paste( input$name,input$surname, Sys.Date(), ".pdf", sep = "")
        },
        content = function(file) {
          # Create the R Markdown content
          rmd_content <- sprintf('RAREINSIGHT REPORT
                                 
                            User Information : 
                            __________________
                                  
                            Name: %s
                            Surname: %s
                            Ethnicity: %s
                            Date of Birth: %s
                            Clinical Diagnosis (OMIM): %s
                            Phenotype (HPO Terms): %s
                            Test Performed:  %s',
                            input$name, input$surname, input$ethnicity, input$dob, input$clinical_diagnosis, input$phenotype, input$test_performed)
                            img(src = "rareinsight_final.png", height = 50)
          
          # Create a temporary file to store the R Markdown content
          temp_rmd_file <- tempfile(fileext = ".Rmd")
          # Write the R Markdown content to the temporary file
          writeLines(rmd_content, temp_rmd_file)
          
          # Render the R Markdown file to PDF
          rmarkdown::render(temp_rmd_file, output_file = file)
          # Remove the temporary R Markdown file
          unlink(temp_rmd_file)
        },
        contentType = "application/pdf"  # Set content type to PDF
  )
  
  ### Search Panel part
  observeEvent(input$search_button, {
    search_term <- input$search_input
    search_type <- input$search_type
    # Define the column to search based on search type
    search_column <- switch(
      search_type,
      "Gene" = "GeneSymbol",
      "Variant" = "Name",
      "Disorder" = "PhenotypeList",
      "dbSNP" = "RS...dbSNP.",
      "ClinVar" = "RCVaccession"
    )
    # Filter data based on search term and column
    filtered_data <- variant_summary()[grep(search_term, variant_summary()[[search_column]], ignore.case = TRUE), ]
    output$search_results <- renderDT({
      if(nrow(filtered_data) == 0) {
        # Display a message if no results are found
        HTML("<p>No matching results found.</p>")
      } else {
        datatable(filtered_data, options = list(scrollX = TRUE))
      }
    })
  })
  
  ### buttonOMIM
  observeEvent(input$buttonOMIM, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the OMIM search URL with the search term and open it in the browser
      omim_search_url <- paste0("https://www.omim.org/search?index=entry&start=1&search=", URLencode(search_term), "&sort=score+desc%2C+prefix_sort+desc&limit=10&date_created_from=&date_created_to=&date_updated_from=&date_updated_to=")
      browseURL(omim_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  ### buttonClinVar 
  observeEvent(input$buttonClinVar, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the Clinvar search URL with the search term and open it in the browser
      ClinVar_search_url <- paste0("https://www.ncbi.nlm.nih.gov/clinvar/?term=", URLencode(search_term), "%5Bgene%5D&redir=gene")
      browseURL(ClinVar_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  ### buttonPubMed 
  observeEvent(input$buttonPubMed, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the Clinvar search URL with the search term and open it in the browser
      PubMed_search_url <- paste0("https://pubmed.ncbi.nlm.nih.gov/?term=", URLencode(search_term))
      browseURL(PubMed_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  ### buttonGeneReviews 
  observeEvent(input$buttonGeneReviews, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the Clinvar search URL with the search term and open it in the browser
      GeneReviews_search_url <- paste0("https://www.ncbi.nlm.nih.gov/books/NBK1116/?term=", URLencode(search_term), "gene%20review")
      browseURL(GeneReviews_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  ### buttonLitVarsome
  observeEvent(input$buttonVarsome, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the Clinvar search URL with the search term and open it in the browser
      Varsome_search_url <- paste0("https://varsome.com/gene/hg38/", URLencode(search_term))
      browseURL(Varsome_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  ### buttonLitVar 
  observeEvent(input$buttonLitVar, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the Clinvar search URL with the search term and open it in the browser
      LitVar_search_url <- paste0("https://www.ncbi.nlm.nih.gov/research/litvar2/docsum?text=", URLencode(search_term))
      browseURL(LitVar_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  ### buttonLdbSNP 
  observeEvent(input$buttondbSNP, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, construct the Clinvar search URL with the search term and open it in the browser
      dbSNP_search_url <- paste0("https://www.ncbi.nlm.nih.gov/snp/?term=", URLencode(search_term))
      browseURL(dbSNP_search_url)
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  observeEvent(input$buttongenomAD, {
    search_term <- isolate(input$search_input)  # Get the search term from the Search Panel
    if (nchar(search_term) > 0) {
      # If there's a search term, query Ensembl REST API to get Ensembl ID
      ensembl_api_url <- paste0("https://rest.ensembl.org/lookup/symbol/human/", URLencode(search_term), "?content-type=application/json")
      ensembl_response <- httr::GET(ensembl_api_url)
      if (httr::http_error(ensembl_response)) {
        # If there's an error in API response, show an alert message
        showModal(
          modalDialog(
            title = "Error",
            "There was an error retrieving data from Ensembl.",
            easyClose = TRUE
          )
        )
      } else {
        ensembl_data <- httr::content(ensembl_response, as = "parsed")
        if (length(ensembl_data) > 0) {
          # If Ensembl ID found, construct the gnomAD search URL with the Ensembl ID and open it in the browser
          ensembl_id <- ensembl_data$id
          gnomad_search_url <- paste0("https://gnomad.broadinstitute.org/gene/", ensembl_id)
          browseURL(gnomad_search_url)
        } else {
          # If Ensembl ID not found, show an alert message
          showModal(
            modalDialog(
              title = "Error",
              "No matching gene found in Ensembl.",
              easyClose = TRUE
            )
          )
        }
      }
    } else {
      # If there's no search term, show an alert message
      showModal(
        modalDialog(
          title = "Error",
          "Please enter a search term in the Search Panel first.",
          easyClose = TRUE
        )
      )
    }
  })
  
  ### buttonbuttonRareConnect 
  observeEvent(input$buttonRareConnect, {
      browseURL("https://www.rareconnect.org/en")
  })
  ### buttonbuttonRAReSOURCE
  observeEvent(input$buttonRAReSOURCE, {
    browseURL("https://raresource.nih.gov/diseases")
  })
   
  ## VCF upload file part
  vcf_data <- reactive({
    req(input$input_file)
    vcf_file <- input$input_file$datapath
    vcf_data <- read.vcfR(vcf_file)
    return(vcf_data)
  })

  output$variant_table <- renderTable({
    req(vcf_data())

    # Extract variant information and filter PASS variants
    variant_info <- as.data.frame(vcf_data()@fix)
    ############ variant_info <- variant_info[variant_info$FILTER == "PASS", ]

    # Remove ID and QUAL columns
    ############ variant_info <- subset(variant_info, select = -c(ID, QUAL))

    # Remobe ID is not Abvailable "NA"
    if (any(is.na(variant_info$ID))) {
      variant_info <- subset(variant_info, select = -c(ID))
    }

    # Split INFO column into separate columns
    info_split <- strsplit(as.character(variant_info$INFO), ";")
    max_splits <- max(lengths(info_split))

    # Create dataframe to store split columns
    info_df <- data.frame(matrix(NA, ncol = max_splits, nrow = nrow(variant_info)))

    # Extract column names and values from INFO column
    info_names_values <- lapply(info_split, function(x) {
      splits <- strsplit(x, "=")
      names <- sapply(splits, `[`, 1)
      values <- sapply(splits, `[`, 2)
      return(list(names = names, values = values))
    })

    # Fill in split columns with appropriate column names
    for (i in 1:max_splits) {
      column_values <- sapply(info_names_values, function(info) ifelse(length(info$values) >= i, info$values[i], NA))
      col_names <- unique(unlist(sapply(info_names_values, function(info) ifelse(length(info$names) >= i, info$names[i], NA))))
      colnames(info_df)[i] <- col_names
      info_df[, i] <- column_values
    }

    # Combine split INFO columns with original dataframe
    variant_info <- cbind(variant_info, info_df)

    # Remove original INFO column
    variant_info <- subset(variant_info, select = -INFO)

    return(variant_info)
  })
 
  
  # Generate  PLOT2
  output$PLOT1 <- renderPlot({
    vcf <- vcf_data()
    plot(vcf)
    chrom <- create.chromR(name='Supercontig', vcf=vcf)
    plot(chrom)
    chrom <- proc.chromR(chrom, verbose=TRUE)
    plot(chrom)
  }) 
  
  # Generate PLOT2
  output$PLOT2 <- renderPlot({
    vcf <- vcf_data()
    plot(vcf)
    chrom <- create.chromR(name='Supercontig', vcf=vcf)
    plot(chrom)
    chrom <- proc.chromR(chrom, verbose=TRUE)
    plot(chrom)
    chromoqc(chrom, dp.alpha=20)
    #      chromoqc(chrom, xlim=c(5e+05, 6e+05))
    
  })
}

shinyApp(ui, server)

                                 
